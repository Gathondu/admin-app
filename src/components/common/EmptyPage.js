import React from "react";
import PropTypes from "prop-types";

const EmptyPage = ({text}) => {
  return (
    <div className="jumbotron">
      <i className="fa fa-ban fa-5x" aria-hidden="true" />
      <h2>{text}</h2>
    </div>
  );
};

  EmptyPage.propTypes = {
      text: PropTypes.string.isRequired
  };

  EmptyPage.defaultProps = {
      text: "No Courses to display. Add course to view"
  };

export default EmptyPage;
