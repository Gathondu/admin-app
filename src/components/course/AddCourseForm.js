import React from 'react';
import PropTypes from 'prop-types';

let courseTitle = React.createRef();
let courseDescription = React.createRef();
const AddCourseForm = ({addCourse}) => {
    let handleSubmit = (e) => {
        e.preventDefault();
        addCourse(courseTitle.current.value, courseDescription.current.value);
        e.currentTarget.reset();
    };

    return (
        <>
        <h2>Add Course</h2>
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                placeholder="Course Title"
                ref={courseTitle} />
            <input
                type="text"
                placeholder="Course Description"
                ref={courseDescription} />
            <input
                type="submit"
                value="Save"/>
        </form>
        </>
    );
};

AddCourseForm.propTypes = {
    addCourse: PropTypes.func.isRequired
};

export default AddCourseForm;
