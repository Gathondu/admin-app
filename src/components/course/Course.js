import React from 'react';
import PropTypes from 'prop-types';

const Course = ({title, description, index, removeCourse}) => {
    return (
        <div className="course">
            <span className="course-title">
                <button className="remove-course" onClick={() => removeCourse(index)}>Ã—</button>
                {title}
                <br/>
                {description}
            </span>
        </div>
    );
};

Course.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    removeCourse: PropTypes.func.isRequired
};

export default Course;
