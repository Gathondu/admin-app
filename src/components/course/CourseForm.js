import React from 'react';
import PropTypes from 'prop-types';
import TextInput from '../common/TextInput'
import SelectInput from '../common/SelectInput'

const CourseForm = ({course, allAuthors, onSave, onChange, loading, errors, formIsValid}) => {
    return (
        <form>
            <h1>Manage Course</h1>
            <TextInput
                label="Title"
                onChange={onChange}
                name="title"
                value={course.title}
                error={errors.title}
                placeholder="Enter Course Title"/>
            <SelectInput
                label="Author"
                onChange={onChange}
                name="authorId"
                value={course.authorId}
                defaultOption="Select Author"
                options={allAuthors}
                error={errors.authorId}/>
            <TextInput
                label="Category"
                onChange={onChange}
                name="category"
                value={course.category}
                error={errors.category}
                placeholder="Enter Course Category"/>
            <TextInput
                label="Length"
                onChange={onChange}
                name="length"
                value={course.length}
                error={errors.length}
                placeholder="Enter Course Length"/>
            <input
                type="submit"
                disabled={loading || !formIsValid}
                value={loading ? 'Saving...' : 'Save'}
                className="btn btn-primary"
                onClick={onSave}/>
        </form>
    );
};

CourseForm.propTypes = {
    course: PropTypes.object.isRequired,
    allAuthors: PropTypes.array,
    onSave: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    loading: PropTypes.bool,
    errors: PropTypes.object
};
export default CourseForm;
