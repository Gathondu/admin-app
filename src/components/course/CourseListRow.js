import React from 'react';
import PropTypes from 'prop-types';
import {Link} from "react-router-dom";

const CourseListRow = ({course, deleteCourse}) => {
    return (
        <tr>
            <td><i className="fa fa-trash" onClick={() => deleteCourse(course.id)}/></td>
            <td><a href={course.watchHref} target="_blank" rel="noopener noreferrer"><i className="fa fa-television"/></a></td>
            <td><Link to={`/course/${course.id}`}>{course.title}</Link></td>
            <td>{course.authorId}</td>
            <td>{course.category}</td>
            <td>{course.length}</td>
        </tr>
    );
};

CourseListRow.propTypes = {
    course: PropTypes.object.isRequired
};

export default CourseListRow;
