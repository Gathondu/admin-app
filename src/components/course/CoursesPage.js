import React, {Fragment} from "react";
import PropTypes from "prop-types";
import CourseList from "./CourseList";

const CoursesPage = ({ history }) => {
    return <Fragment>
      <h1 key="title">Courses</h1>
      <input
        key="submit"
        type="submit"
        value="Add Course"
        className="btn btn-primary"
        onClick={() => history.push("/course")}
      />
      <CourseList key="course-list" />
    </Fragment>
  };

CoursesPage.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default CoursesPage;
