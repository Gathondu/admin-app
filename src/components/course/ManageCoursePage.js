import React, { Component } from "react";
import { connect } from "react-redux";
import { Prompt } from "react-router-dom";
import PropTypes from "prop-types";
import { bindActionCreators } from "redux";
import toastr from "toastr";
import * as courseActions from "../../actions/course";
import CourseForm from "./CourseForm";

class ManageCoursePage extends Component {
  static propTypes = {
    course: PropTypes.object.isRequired,
    authors: PropTypes.array.isRequired,
    actions: PropTypes.object.isRequired
  };
  state = {
    course: Object.assign({}, this.props.course),
    loading: false,
    errors: { title: "", author: "", category: "", length: "" },
    formIsValid: false,
    titleValid: false,
    authorValid: false,
    categoryValid: false,
    lengthValid: false,
    formUpdated: false
  };
  componentWillReceiveProps(nextProps) {
    if (this.props.course.id !== nextProps.course.id) {
      this.setState({ course: Object.assign({}, nextProps.course) });
    }
  }
  shouldComponentUpdate(nextProps, nextState, nextContext) {
    if (this.state.course !== nextState.course) {
      this.setState({ formUpdated: true });
    }
    return true;
  }

  validateInput = course => {
    let errors = {};
    for (var field in course) {
      if (course.hasOwnProperty(field)) {
        switch (field) {
          case "title":
            if (course["title"].length > 0) {
              this.setState({ titleValid: true });
            } else {
              errors["title"] = "Title should contain more than 1 character";
            }
            break;
          case "authorId":
            if (course["authorId"] !== "") {
              this.setState({ authorValid: true });
            } else {
              errors["authorId"] = "Please select one of the authors";
            }
            break;
          case "category":
            if (course["category"].length > 0) {
              this.setState({ categoryValid: true });
            } else {
              errors["category"] =
                "Category should contain more than 1 character";
            }
            break;
          case "length":
            if (
              course["length"].length > 0 &&
              course["length"].match(/\[a-zA-z]+/g) === null &&
              course["length"].match(/^(\d{1,2}):(\d{1,2})/g) !== null
            ) {
              this.setState({ lengthValid: true });
            } else {
              errors["length"] =
                "Length should neither be empty nor contain words. Just numbers and a colon in the format 04:40";
            }
            break;
          default:
            break;
        }
      }
    }
    if (Object.keys(errors).length > 0) {
      this.setState({ errors: errors });
    } else {
      this.setState({ errors: {}, formIsValid: true });
    }
  };

  updateCoursesState = e => {
    const field = e.target.name;
    let course = Object.assign({}, this.state.course);
    course[field] = e.target.value;
    this.validateInput(course);
    this.setState({ course: course });
  };

  saveCourse = e => {
    e.preventDefault();
    this.setState({ loading: true });
    this.validateInput(this.state.course);
    if (this.state.formIsValid) {
      this.setState({ formUpdated: false });
      this.props.actions
        .saveCourse(this.state.course)
        .then(() => this.redirect())
        .catch(error => {
          this.setState({ loading: false });
          toastr.error(error);
        });
    }
  };

  redirect = () => {
    this.setState({ loading: false });
    toastr.success("Course Saved.");
    this.props.history.push("/courses");
  };

  render() {
    return [
      <Prompt
        key="prompt"
        when={this.state.formUpdated}
        message="The form isn't saved, are you sure you want to leave the page?"
      />,
      <CourseForm
        key="form"
        course={this.state.course}
        errors={this.state.errors}
        allAuthors={this.props.authors}
        onChange={this.updateCoursesState}
        onSave={this.saveCourse}
        loading={this.state.loading}
        formIsValid={this.state.formIsValid}
      />
    ];
  }
}

const getCourseById = (courses, id) => {
  const course = courses.filter(course => course.id === id);
  if (course.length) return course[0];
  return null;
};
const mapStateToProps = (state, ownProps) => {
  const courseId = ownProps.match.params.id; // get this from path `/course/:id`
  let course = {
    id: "",
    watchHref: "",
    title: "",
    authorId: "",
    length: "",
    category: ""
  };
  if (courseId && state.courses.length > 0) {
    course = getCourseById(state.courses, courseId);
  }
  const authorsFormattedForDropdown = state.authors.map(author => {
    return {
      value: author.id,
      text: author.firstName + " " + author.lastName
    };
  });

  return {
    course: course,
    authors: authorsFormattedForDropdown
  };
};
const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(courseActions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ManageCoursePage);
