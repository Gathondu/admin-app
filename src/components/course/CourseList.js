import React from "react";
import PropTypes from "prop-types";
import CourseListRow from "./CourseListRow";
import LoadingDots from '../common/LoadingDots';
import EmptyPage from '../common/EmptyPage';
import { connect } from 'react-redux';
import {deleteCourse} from '../../actions/course';

export const CourseList = ({ courses, deleteCourse, loading }) => {
  const courseRows = courses.map((course, index) => (
    <CourseListRow
      course={course}
      key={course.id}
      index={index}
      deleteCourse={deleteCourse}
    />
  ));
  if (courseRows.length > 0) {
    return (
      <table className="table">
        <thead>
          <tr>
            <th>&nbsp;</th>
            <th>&nbsp;</th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
            <th>Length</th>
          </tr>
        </thead>
        <tbody>{courseRows}</tbody>
      </table>
    );
  } else if (!loading) {
    return <EmptyPage />;
  } else {
      return <h3>Loading <LoadingDots interval={100} dots={20} /></h3>;
  }
};

CourseList.propTypes = {
  courses: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  deleteCourse: PropTypes.func.isRequired
};
export const mapStateToProps = state => (
    {
      courses: state.courses,
      loading: state.ajaxCallInProgress > 0
    }
);

export default connect(mapStateToProps, {deleteCourse})(CourseList);
