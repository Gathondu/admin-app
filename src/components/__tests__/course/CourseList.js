import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {CourseList, mapStateToProps} from "../../course/CourseList";
import EmptyPage from "../../common/EmptyPage";
import LoadingDots from "../../common/LoadingDots";
import CourseListRow from "../../course/CourseListRow";
import {courses} from "../../../api/mockCourse";

describe('CourseList Component', () => {
    const setup = () => {
        const props = {
            courses: courses,
            loading: false,
            deleteCourse: jest.fn()
        };

        return {
            props,
            presentationWrapper: shallow(<CourseList { ...props } />)
        };
    };
    const setupFalsy = () => {
        const props = {
            courses: [],
            loading: true,
            deleteCourse: jest.fn()
        };

        return {
            props,
            presentationWrapper: shallow(<CourseList { ...props } />)
        };
    };
    const setupEmpty = () => {
        const props = {
            courses: [],
            loading: false,
            deleteCourse: jest.fn()
        };

        return {
            props,
            presentationWrapper: shallow(<CourseList { ...props } />)
        };
    };
    it('should render correctly', () => {
        const {presentationWrapper} = setup();
        expect(toJson(presentationWrapper)).toMatchSnapshot();
    });
    it('should render courses list correctly', () => {
        const {presentationWrapper} = setup();
        expect(presentationWrapper.find(CourseListRow)).toHaveLength(5);
    });
    it('should load state correctly', () => {
        const {props} = setup();
        expect(mapStateToProps(props).courses).toHaveLength(5);
    });
    it('should show loading dots if loading is true', () => {
        const {presentationWrapper} = setupFalsy();
        expect(presentationWrapper.find(LoadingDots)).toHaveLength(1);
    });
    it('should show empty page if no courses to display', () => {
        const {presentationWrapper} = setupEmpty();
        expect(presentationWrapper.find(EmptyPage)).toHaveLength(1);
    });
});