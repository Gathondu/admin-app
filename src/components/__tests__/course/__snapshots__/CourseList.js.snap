// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseList Component should render correctly 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
         
      </th>
      <th>
         
      </th>
      <th>
        Title
      </th>
      <th>
        Author
      </th>
      <th>
        Category
      </th>
      <th>
        Length
      </th>
    </tr>
  </thead>
  <tbody>
    <CourseListRow
      course={
        Object {
          "authorId": "cory-house",
          "category": "JavaScript",
          "id": "react-flux-building-applications",
          "length": "5:08",
          "title": "Building Applications in React and Flux",
          "watchHref": "http://www.pluralsight.com/courses/react-flux-building-applications",
        }
      }
      deleteCourse={[MockFunction]}
      index={0}
      key="react-flux-building-applications"
    />
    <CourseListRow
      course={
        Object {
          "authorId": "cory-house",
          "category": "Software Practices",
          "id": "clean-code",
          "length": "3:10",
          "title": "Clean Code: Writing Code for Humans",
          "watchHref": "http://www.pluralsight.com/courses/writing-clean-code-humans",
        }
      }
      deleteCourse={[MockFunction]}
      index={1}
      key="clean-code"
    />
    <CourseListRow
      course={
        Object {
          "authorId": "cory-house",
          "category": "Software Architecture",
          "id": "architecture",
          "length": "2:52",
          "title": "Architecting Applications for the Real World",
          "watchHref": "http://www.pluralsight.com/courses/architecting-applications-dotnet",
        }
      }
      deleteCourse={[MockFunction]}
      index={2}
      key="architecture"
    />
    <CourseListRow
      course={
        Object {
          "authorId": "cory-house",
          "category": "Career",
          "id": "career-reboot-for-developer-mind",
          "length": "2:30",
          "title": "Becoming an Outlier: Reprogramming the Developer Mind",
          "watchHref": "http://www.pluralsight.com/courses/career-reboot-for-developer-mind",
        }
      }
      deleteCourse={[MockFunction]}
      index={3}
      key="career-reboot-for-developer-mind"
    />
    <CourseListRow
      course={
        Object {
          "authorId": "cory-house",
          "category": "HTML5",
          "id": "web-components-shadow-dom",
          "length": "5:10",
          "title": "Web Component Fundamentals",
          "watchHref": "http://www.pluralsight.com/courses/web-components-shadow-dom",
        }
      }
      deleteCourse={[MockFunction]}
      index={4}
      key="web-components-shadow-dom"
    />
  </tbody>
</table>
`;
